### PR TITLE
fix: guard inner index binary deserialization

### LINKED ISSUE #1964: Guard InnerIndexInterface deserialize reads against empty binary data
Parent tracking issue: #1960

The x86 ASan/UBSan functest job reports a null pointer passed to `memcpy` during `InnerIndexInterface::Deserialize`.

CI evidence:

- Test X86 Functests: https://github.com/antgroup/vsag/actions/runs/25100797892/job/73553696201

Log excerpt:

```text
/__w/vsag/vsag/src/algorithm/inner_index_interface.cpp:200:20: runtime error: null pointer passed as argument 1, which is declared to never be null
```

Code location:

- `src/algorithm/inner_index_interface.cpp:197-200` obtains `Binary b = binary_set.Get(this->GetName())`, then the read lambda does `std::memcpy(dest, b.data.get() + offset, len)`.

Initial root cause:

The deserialize path assumes the requested binary entry exists and has valid data. If the key is missing, the binary data is empty/null, or a zero/invalid read is routed through the callback, the lambda can pass null pointers into `memcpy` or perform pointer arithmetic on null data.

Suggested fix:

- Validate that the binary set contains the expected key before constructing the reader.
- Validate `b.data != nullptr` when `b.size > 0`.
- Make the read callback return immediately for `len == 0`.
- Check `offset + len <= b.size` before reading and throw a `READ_ERROR` or equivalent existing error type on invalid input.
- Add regression tests for missing/empty binary entries and invalid read bounds.

Non-goals:

- Do not change serialized format.
- Do not silently accept malformed serialized data unless the existing API contract requires it.
- Do not mix bitset/HNSW fixes into this PR.

### PR DESCRIPTION
**Pull request by jac0626**
## Summary

Guard `InnerIndexInterface::Deserialize(const BinarySet&)` against malformed binary input before constructing the read callback.

The binary-set path now reports `READ_ERROR` for missing index data, null non-empty data, null read destinations, and out-of-range reads. Regression coverage exercises missing, null, and truncated index binary entries.

Fixes #1964.
Part of #1960.

## Validation

- `git diff --check`
- Fork PR CI: https://github.com/jac0626/vsag/actions/runs/25477024729
- Log scan: target `inner_index_interface.cpp` runtime error is gone in this branch; remaining runtime errors are existing issues tracked separately (`fast_bitset.cpp`, `sparse_term_datacell.cpp`, `hnswalg.*`).

## Notes

Local macOS rebuild was blocked by the existing CMake configuration error: `gcc does not support using libc++`.

### CODE REVIEW

File: src/algorithm/inner_index_interface.cpp [hunk 1]
```diff
@@ -194,9 +196,29 @@ InnerIndexInterface::Deserialize(const BinarySet& binary_set) {
...
+        if (dest == nullptr) {
+            throw VsagException(
+                ErrorType::READ_ERROR, "null read destination for index: ", this->GetName());
+        }
+        if (b.data == nullptr || offset > b.size || len > b.size - offset) {
+            throw VsagException(
+                ErrorType::READ_ERROR, "binary read out of range for index: ", this->GetName());
+        }
         std::memcpy(dest, b.data.get() + offset, len);
     };
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/algorithm/inner_index_interface.cpp:245]: Fixed in `5929bc19`. The callback now checks `offset` and `len` against `std::numeric_limits<size_t>::max()` before casting and uses the checked `size_t` values for pointer arithmetic and `memcpy`.

File: src/algorithm/inner_index_interface_test.cpp [hunk 2]
```diff
@@ -123,6 +125,32 @@ class EmptyInnerIndex : public InnerIndexInterface {
...
+RequireReadError(const std::function<void()>& func) {
+    bool got_read_error = false;
+    try {
+        func();
+    } catch (const VsagException& e) {
+        got_read_error = true;
+        REQUIRE(e.error_.type == ErrorType::READ_ERROR);
+    }
+    REQUIRE(got_read_error);
+}
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/algorithm/inner_index_interface_test.cpp:194]: Fixed in `5929bc19`. `RequireReadError` is now a function template, so the call sites stay the same without `std::function` type erasure.

File: src/algorithm/inner_index_interface.cpp [hunk 3]
```diff
@@ -194,9 +196,29 @@ InnerIndexInterface::Deserialize(const BinarySet& binary_set) {
...
+            return;
+        }
+        if (dest == nullptr) {
+            throw VsagException(
+                ErrorType::READ_ERROR, "null read destination for index: ", this->GetName());
+        }
+        if (b.data == nullptr || offset > b.size || len > b.size - offset) {
+            throw VsagException(
+                ErrorType::READ_ERROR, "binary read out of range for index: ", this->GetName());
+        }
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/algorithm/inner_index_interface.cpp:231]: Fixed in `5929bc19`. Added targeted regression coverage for both `dest == nullptr` and seek/read out-of-range callback validation paths.

File: src/algorithm/inner_index_interface.cpp [hunk 4]
```diff
@@ -194,10 +197,37 @@ InnerIndexInterface::Deserialize(const BinarySet& binary_set) {
...
+        if (offset > std::numeric_limits<size_t>::max() ||
+            len > std::numeric_limits<size_t>::max()) {
+            throw VsagException(
+                ErrorType::READ_ERROR, "binary read too large for index: ", this->GetName());
+        }
+        const auto copy_offset = static_cast<size_t>(offset);
+        const auto copy_len = static_cast<size_t>(len);
+        std::memcpy(dest, b.data.get() + copy_offset, copy_len);
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/algorithm/inner_index_interface.cpp]: Fixed in `40000b9a`. Added `std::ptrdiff_t` range validation before pointer arithmetic and cast the checked offset through `std::ptrdiff_t`.

File: src/algorithm/inner_index_interface_test.cpp [hunk 5]
```diff
@@ -176,3 +233,46 @@ TEST_CASE("InnerIndexInterface NOT Implemented", "[ut][InnerIndexInterface]") {
...
+
+    SECTION("missing index binary") {
+        BinarySet binary;
+
+        RequireReadError([&index, &binary]() { index->Deserialize(binary); });
+    }
+
+    SECTION("null non-empty index binary") {
+        BinarySet binary;
+        binary.Set(index->GetName(), Binary{.data = nullptr, .size = sizeof(uint64_t)});
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/algorithm/inner_index_interface_test.cpp]: Fixed in `40000b9a`. Replaced C++20 designated initializers with explicit `Binary` field assignment via a small test helper.

File: src/algorithm/inner_index_interface_test.cpp [hunk 6]
```diff
@@ -123,6 +125,61 @@ class EmptyInnerIndex : public InnerIndexInterface {
...
+RequireReadError(Func&& func) {
+    bool got_read_error = false;
+    try {
+        std::forward<Func>(func)();
+    } catch (const VsagException& e) {
+        got_read_error = true;
+        REQUIRE(e.error_.type == ErrorType::READ_ERROR);
+    }
+    REQUIRE(got_read_error);
+}
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/algorithm/inner_index_interface_test.cpp:194]: Fixed in `40000b9a`. `RequireReadError` now catches unexpected `std::exception` and non-standard exceptions with explicit test failures.

File: src/algorithm/inner_index_interface.cpp [hunk 7]
```diff
@@ -194,10 +199,47 @@ InnerIndexInterface::Deserialize(const BinarySet& binary_set) {
...
         // logger::debug("read offset {} len {}", offset, len);
-        std::memcpy(dest, b.data.get() + offset, len);
+        if (len == 0) {
+            return;
+        }
+        if (dest == nullptr) {
+            throw VsagException(
+                ErrorType::READ_ERROR, "null read destination for index: ", index_name);
+        }
+        if (b.data == nullptr || offset > b.size || len > b.size - offset) {
```
Comment (LHT129) [inline src/algorithm/inner_index_interface.cpp:228]: Core bounds check is correct: because offset > b.size is evaluated first (short-circuit), the subsequent len > b.size - offset cannot underflow, so no unsigned-wraparound OOB slips through. This is the right guard for the memcpy.

File: src/algorithm/inner_index_interface.cpp [hunk 8]
```diff
@@ -194,10 +199,47 @@ InnerIndexInterface::Deserialize(const BinarySet& binary_set) {
...
+            throw VsagException(
+                ErrorType::READ_ERROR, "binary read out of range for index: ", index_name);
+        }
+        if (len > std::numeric_limits<size_t>::max()) {
```
Comment (LHT129) [inline src/algorithm/inner_index_interface.cpp:232]: len > std::numeric_limits<size_t>::max() is dead code on the supported 64-bit targets: len is uint64_t and size_t is also 64-bit, so this can never be true. Harmless, but it (and the static_cast to size_t) only does anything on a hypothetical 32-bit build. Consider dropping it or documenting the intent, since the ptrdiff_t check below is the one that actually constrains the value range.

File: src/algorithm/inner_index_interface.cpp [hunk 9]
```diff
@@ -194,10 +199,47 @@ InnerIndexInterface::Deserialize(const BinarySet& binary_set) {
...
+    if (b.size > 0 && b.data == nullptr) {
+        throw VsagException(ErrorType::READ_ERROR, "null binary data for index: ", index_name);
+    }
+
+    const auto* binary_data = b.data.get();
+    using PointerDiffLimit = std::make_unsigned_t<std::ptrdiff_t>;
+    const auto max_pointer_offset =
+        static_cast<PointerDiffLimit>(std::numeric_limits<std::ptrdiff_t>::max());
+
     auto func = [&](uint64_t offset, uint64_t len, void* dest) -> void {
```
Comment (LHT129) [inline src/algorithm/inner_index_interface.cpp:219]: Scope note: this guard protects only the memcpy in the BinarySet path. The concrete Deserialize(StreamReader&) impls (hgraph/brute_force) call ReadVector/ReadString which resize(size) on an attacker-controlled length before any read, so a crafted binary can still raise std::bad_alloc here. Since VsagException/std::bad_alloc are not std::runtime_error, the existing catch (const std::runtime_error&) block does not convert them — bad_alloc would propagate unwrapped instead of becoming NO_ENOUGH_MEMORY. Not a regression vs prior behavior, but the deserialization surface is not fully hardened against resource-exhaustion DoS.

**Review by LHT129 — COMMENTED** [review]
The hardening of the BinarySet deserialization path is sound. The core bounds check correctly avoids unsigned underflow (offset > b.size is tested before b.size - offset), null pointers are rejected for both data and dest, len==0 short-circuits cleanly, and ptrdiff_t representability is validated before the memcpy, so a crafted Binary can no longer trigger an OOB read in this lambda. Errors are raised as VsagException(READ_ERROR), which is the correct failure mode and well covered by the new tests. Two notes worth considering: (1) the `len > numeric_limits<size_t>::max()` check is dead on the only supported 64-bit platforms (size_t == uint64_t == len's type), and (2) the guard only covers the BinarySet path's memcpy; the actual Deserialize(StreamReader&) implementations still resize() vectors/strings from attacker-controlled lengths, so a crafted blob can still throw std::bad_alloc before any read is bounds-checked. Since VsagException derives from std::exception (not std::runtime_error), a bad_alloc would surface unwrapped rather than as NO_ENOUGH_MEMORY. None of these are regressions; the primary OOB-read vulnerability is fixed.

File: src/algorithm/inner_index_interface.cpp [hunk 10]
```diff
@@ -194,10 +199,45 @@ InnerIndexInterface::Deserialize(const BinarySet& binary_set) {
...
+                ErrorType::READ_ERROR, "binary read out of range for index: ", this->GetName());
+        }
+        if (offset > std::numeric_limits<size_t>::max() ||
+            len > std::numeric_limits<size_t>::max()) {
+            throw VsagException(
+                ErrorType::READ_ERROR, "binary read too large for index: ", this->GetName());
+        }
+        const auto copy_offset = static_cast<size_t>(offset);
+        const auto copy_len = static_cast<size_t>(len);
+        using PointerDiffLimit = std::make_unsigned_t<std::ptrdiff_t>;
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/algorithm/inner_index_interface.cpp]: Fixed in `40000b9a`. `copy_offset` is now used for the final pointer arithmetic on line 244: `binary_data + copy_offset`.

File: src/algorithm/inner_index_interface.cpp [hunk 11]
```diff
@@ -194,10 +199,45 @@ InnerIndexInterface::Deserialize(const BinarySet& binary_set) {
...
+        if (offset > static_cast<PointerDiffLimit>(std::numeric_limits<std::ptrdiff_t>::max()) ||
+            len > static_cast<PointerDiffLimit>(std::numeric_limits<std::ptrdiff_t>::max()) ||
+            len > static_cast<PointerDiffLimit>(std::numeric_limits<std::ptrdiff_t>::max()) -
+                      offset) {
+            throw VsagException(
+                ErrorType::READ_ERROR, "binary read offset too large for index: ", this->GetName());
+        }
+        std::memcpy(dest, b.data.get() + static_cast<std::ptrdiff_t>(offset), copy_len);
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/algorithm/inner_index_interface.cpp]: Fixed in `40000b9a`. The callback now validates `offset`/`len` once (ptrdiff_t representability), then uses the casted values (`copy_offset`, `copy_len`) consistently for both pointer arithmetic and memcpy length.

File: src/algorithm/inner_index_interface.cpp [hunk 12]
```diff
@@ -194,10 +199,44 @@ InnerIndexInterface::Deserialize(const BinarySet& binary_set) {
...
 
+    if (not binary_set.Contains(this->GetName())) {
+        throw VsagException(
+            ErrorType::READ_ERROR, "missing binary data for index: ", this->GetName());
+    }
+
     Binary b = binary_set.Get(this->GetName());
+    if (b.size > 0 && b.data == nullptr) {
+        throw VsagException(ErrorType::READ_ERROR, "null binary data for index: ", this->GetName());
+    }
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/algorithm/inner_index_interface.cpp]: Already addressed in `4684b00a`. `GetName()` is cached as `const auto index_name = this->GetName();` on line 204 and reused throughout.

File: src/algorithm/inner_index_interface_test.cpp [hunk 13]
```diff
@@ -176,3 +245,46 @@ TEST_CASE("InnerIndexInterface NOT Implemented", "[ut][InnerIndexInterface]") {
...
+        BinarySet binary;
+        binary.Set(index->GetName(), MakeBinary(std::shared_ptr<int8_t[]>(), sizeof(uint64_t)));
+
+        RequireReadError([&index, &binary]() { index->Deserialize(binary); });
+    }
+
+    SECTION("truncated index binary") {
+        auto data = std::shared_ptr<int8_t[]>(new int8_t[1]);
+        BinarySet binary;
+        binary.Set(index->GetName(), MakeBinary(data, 1));
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/algorithm/inner_index_interface_test.cpp]: Fixed in `4684b00a`. Test buffers now use value-initialization (`new int8_t[1]{}`).

File: src/algorithm/inner_index_interface.cpp [hunk 14]
```diff
@@ -194,10 +199,44 @@ InnerIndexInterface::Deserialize(const BinarySet& binary_set) {
...
+        const auto max_pointer_offset =
+            static_cast<PointerDiffLimit>(std::numeric_limits<std::ptrdiff_t>::max());
+        if (offset > max_pointer_offset || len > max_pointer_offset ||
+            len > max_pointer_offset - offset) {
+            throw VsagException(
+                ErrorType::READ_ERROR, "binary read offset too large for index: ", index_name);
+        }
+        const auto copy_offset = static_cast<std::ptrdiff_t>(offset);
+        std::memcpy(dest, b.data.get() + copy_offset, copy_len);
     };
```
[the remark this answers is not in the document]
Comment (jac0626) [reply src/algorithm/inner_index_interface.cpp:245]: The lambda is 15 lines in a single call site; extracting a separate helper would add indirection without meaningful reuse. The ptrdiff_t comment was added in `4684b00a` to explain the intent. Keeping as-is.

File: src/algorithm/inner_index_interface.cpp (hunk 7 above)
Comment (jac0626) [reply src/algorithm/inner_index_interface.cpp:228]: Thanks for the confirmation on the short-circuit evaluation preventing underflow.

File: src/algorithm/inner_index_interface.cpp (hunk 8 above)
Comment (jac0626) [reply src/algorithm/inner_index_interface.cpp:232]: Agreed it's dead code on our 64-bit targets. Keeping it as a safety net for hypothetical 32-bit builds since it's zero-cost (compiler elides the branch). The ptrdiff_t check is indeed the one that matters in practice.

File: src/algorithm/inner_index_interface.cpp (hunk 9 above)
Comment (jac0626) [reply src/algorithm/inner_index_interface.cpp:219]: Good point. This PR scopes the fix to the BinarySet memcpy path that UBSan reported. The StreamReader-based deserialize paths (ReadVector/ReadString allocations and bad_alloc propagation) are a separate concern and should be tracked in a follow-up issue.

**Review by LHT129 — APPROVED** [review]
LGTM
